File: 2017/03/rs/src/main.rs
use num::complex::Complex;
use std::collections::HashMap;
use std::env;
use std::fs;
use std::time::Instant;

macro_rules! c {
    ($x: expr, $y: expr) => {
        Complex::new($x, $y)
    };
}

fn part1(target: &i32) -> i32 {
    let side = (*target as f64).sqrt().floor() as i32 + 1;
    let mut past_last_square = target - (side - 1).pow(2);
    let half_side = side / 2;
    if past_last_square >= side {
        past_last_square -= side;
    }
    let offset_to_middle = (half_side - past_last_square).abs();
    half_side + offset_to_middle
}

fn get_sum_for_neighbors(
    directions: &Vec<Complex<i32>>,
    grid: &HashMap<Complex<i32>, i32>,
    position: &Complex<i32>,
) -> i32 {
    directions.iter().fold(0i32, |acc, direction| {
        let neighbor = position + direction;
        if grid.contains_key(&neighbor) {
            grid.get(&neighbor).unwrap() + acc
        } else {
            acc
        }
    })
}

fn part2(target: &i32) -> i32 {
    let directions = vec![
        c!(-1, -1), c!(0, -1), c!(1, -1),
        c!(-1,  0),            c!(1,  0),
        c!(-1,  1), c!(0,  1), c!(1,  1),
    ];
    let turn = c!(0, 1);
    let mut grid = HashMap::new();
    let mut position = c!(0, 0);
    grid.insert(position, 1);
    let mut direction = c!(1, 0);
    let mut moves_in_direction = 1;
    loop {
        for _ in 0..2 {
            direction *= turn;
            for _ in 0..moves_in_direction {
                position += direction;
                let new_value = get_sum_for_neighbors(&directions, &grid, &position);
                if new_value > *target {
                    return new_value;
                }
                grid.insert(position, new_value);
            }
        }
        moves_in_direction += 1;
    }
}

fn solve(target: &i32) -> (i32, i32) {
    (part1(target), part2(target))
}

fn get_input(file_path: &String) -> i32 {
    fs::read_to_string(file_path)
        .expect("Error reading input file!")
        .parse()
        .unwrap()
}

fn main() {
    let args: Vec<String> = env::args().collect();
    if args.len() != 2 {
        panic!("Please, add input file path as parameter");
    }
    let now = Instant::now();
    let (part1_result, part2_result) = solve(&get_input(&args[1]));
    println!("P1: {}", part1_result);
    println!("P1: {}", part2_result);
    println!();
    println!("Time: {:7}", now.elapsed().as_secs_f32());
}

File: 2017/03/c/main.c
#include <stdio.h>
#include <stdlib.h>
#include <sys/time.h>
#include <math.h>
#include <complex.h>

typedef double complex Lattice;

typedef int Input;
typedef struct
{
    int part1;
    int part2;
} Results;

int part1(Input targetNumber)
{
    int side = (int)floor(sqrt(targetNumber)) + 1;
    int pastLastSquare = targetNumber - pow((side - 1), 2);
    int halfSide = side / 2;
    if (pastLastSquare >= side)
        pastLastSquare -= side;
    int offsetToMiddle = abs(halfSide - pastLastSquare);
    return halfSide + offsetToMiddle;
}

typedef struct
{
    Lattice key;
    int value;
} GridItem;

typedef struct
{
    GridItem *items;
    int count;
    int size;
} Grid;

void addItem(Grid *grid, Lattice lattice, int value)
{
    if (grid->size == grid->count)
    {
        GridItem *oldItems = grid->items;
        GridItem *newItems = realloc(oldItems, sizeof(GridItem) * (grid->size + 10));
        grid->items = newItems;
        grid->size += 10;
    }
    grid->items[grid->count++] = (GridItem){lattice, value};
}

Grid createGrid()
{
    return (Grid){
        calloc(10, sizeof(GridItem)),
        0,
        10};
}

int getValueForPosition(Grid *grid, Lattice position)
{
    int count = grid->count;
    GridItem *items = grid->items;
    while (count--)
    {
        if (items->key == position)
            return items->value;
        items++;
    }
    return 0;
}

const Lattice DIRECTIONS[8] = {
    -1 - I, -I, 1 - I,
    -1, 1,
    -1 + I, I, 1 + I};

int getSumForNeighbors(Grid *grid, Lattice position)
{
    int total = 0;
    for (int index = 0; index < 8; index++)
        total += getValueForPosition(grid, position + DIRECTIONS[index]);
    return total;
}

int part2(Input target)
{
    Grid grid = createGrid();
    Lattice position = 0;
    Lattice direction = 1;
    addItem(&grid, position, 1);
    int movesInDirection = 1, move, directionMoves, newValue;
    while (1)
    {
        for (move = 0; move < 2; move++)
        {
            direction *= I;
            for (directionMoves = 0; directionMoves < movesInDirection; directionMoves++)
            {
                position += direction;
                newValue = getSumForNeighbors(&grid, position);
                if (newValue > target)
                {
                    free(grid.items);
                    return newValue;
                }
                addItem(&grid, position, newValue);
            }
        }
        movesInDirection++;
    }
}

Results solve(Input input)
{
    return (Results){part1(input), part2(input)};
}

Input getInput(char *filePath)
{
    FILE *file;
    if ((file = fopen(filePath, "r")) == NULL)
    {
        perror("Error reading input file!\n");
        exit(1);
    }
    fseek(file, 0, SEEK_END);
    long length = ftell(file);
    rewind(file);
    char *content = malloc(length);
    fread(content, 1, length, file);
    int targetNumber = atoi(content);
    fclose(file);
    return targetNumber;
}

void freeInput(Input input)
{
}

int main(int argc, char **argv)
{
    if (argc != 2)
    {
        perror("Please, add input file path as parameter");
        exit(1);
    }
    struct timeval starts, ends;
    gettimeofday(&starts, NULL);
    Input input = getInput(argv[1]);
    Results results = solve(input);
    gettimeofday(&ends, NULL);
    freeInput(input);
    printf("P1: %d\n", results.part1);
    printf("P2: %d\n\n", results.part2);
    printf("Time: %.7f\n", (double)((ends.tv_sec - starts.tv_sec) * 1000000 + ends.tv_usec - starts.tv_usec) / 1000000);
    return 0;
}